src/leader/undirected_mesh/mesh.go

```diff
@@ -0,0 +1,283 @@
+			sendMessageMesh(v, s.Forwarders[1], s.PassRight)
+			msg1 := receiveMessageMesh(v, s.Forwarders[0])
+			s.PassRight = msg1
+			msg2 := receiveMessageMesh(v, s.Forwarders[1])
+			s.PassLeft = msg2
+		}
+	}
+	if round == rounds && s.Min == v.GetIndex() && s.Status == candidate {
+		s.Status = leader
+		msg := messageMesh{s.Min, leader}
+		s.Broadcast = &msg
+		s.Leader = v.GetIndex()
+	}
+	setStateMesh(v, &s)
+	return false
+}
+func newMeshState(Id int, State modeType, Neighbours []modeType, Candidates []int, Forwarders []int, Passives []int, PassRight *messageMesh, PassLeft *messageMesh, NeighbourCount []int) *stateMesh {
+	return &stateMesh{
+		Min:                   Id,
+		Status:                State,
+		Neighbours:            Neighbours,
+		Candidates:            Candidates,
+		Forwarders:            Forwarders,
+		Passives:              Passives,
+		PassRight:             PassRight,
+		PassLeft:              PassLeft,
+		Broadcast:             nil,
+		NeighboursToBroadcast: NeighbourCount,
+		Leader:                -1,
+	}
+}
+
+func initializeMesh(v lib.Node) bool {
+	var state modeType
+	var PassRight *messageMesh
+	var PassLeft *messageMesh
+	switch v.GetOutChannelsCount() {
+	case 2:
+		state = candidate
+		for i := 0; i < v.GetOutChannelsCount(); i++ {
+			msg := messageMesh{v.GetIndex(), candidate}
+			sendMessageMesh(v, i, &msg)
+		}
+		PassRight = &messageMesh{v.GetIndex(), candidate}
+		PassLeft = &messageMesh{v.GetIndex(), candidate}
+	case 3:
+		state = forwarding
+		for i := 0; i < v.GetOutChannelsCount(); i++ {
+			msg := messageMesh{v.GetIndex(), forwarding}
+			sendMessageMesh(v, i, &msg)
+		}
+		PassRight = nil
+		PassLeft = nil
+	case 4:
+		state = passive
+		for i := 0; i < v.GetOutChannelsCount(); i++ {
+			msg := messageMesh{v.GetIndex(), passive}
+			sendMessageMesh(v, i, &msg)
+		}
+		PassRight = nil
+		PassLeft = nil
+	}
+	Neighbours, Candidates, Forwarders, Passives := buildNeighboursMesh(v)
+	NeighboursToBroadcast := make([]int, 0)
+	for i := 0; i < v.GetOutChannelsCount(); i++ {
+		NeighboursToBroadcast = append(NeighboursToBroadcast, i)
+	}
+	setStateMesh(v, newMeshState(v.GetIndex(), state, Neighbours, Candidates, Forwarders, Passives, PassRight, PassLeft, NeighboursToBroadcast))
+	return false
+}
+
+func buildNeighboursMesh(v lib.Node) ([]modeType, []int, []int, []int) {
+	Neighbours := make([]modeType, 0)
+	Candidates := make([]int, 0)
+	Forwarders := make([]int, 0)
+	Passives := make([]int, 0)
+	for i := 0; i < v.GetInChannelsCount(); i++ {
+		msg := receiveMessageMesh(v, i)
+		Neighbours = append(Neighbours, msg.Mode)
+		switch msg.Mode {
+		case candidate:
+			Candidates = append(Candidates, i)
+		case forwarding:
+			Forwarders = append(Forwarders, i)
+		case passive:
+			Passives = append(Passives, i)
+		}
+	}
+	return Neighbours, Candidates, Forwarders, Passives
+}
+
+func receiveMessageMesh(v lib.Node, index int) *messageMesh {
+	var msg messageMesh
+	in := v.ReceiveMessage(index)
+	if in != nil {
```

Comment:
FYI: you can do it in one line in Golang: if in := v.ReceiveMessage(index); in != nil { }